pkg/cli/client.go

```diff
@@ -18,20 +18,131 @@ package cli
 
 import (
 	"context"
+	"net/url"
 
 	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/vim25"
 	"github.com/vmware/govmomi/vim25/soap"
 )
 
 type ClientOption struct {
-	hostURL  string
-	insecure bool
+	insecure   bool
+	credential Credential
+	Client     *govmomi.Client
+	url        *url.URL
+	// config Config
+	Config Config
 }
 
-func (o *ClientOption) NewClient(ctx context.Context) (*govmomi.Client, error) {
-	u, err := soap.ParseURL(o.hostURL)
+func (o *ClientOption) NewClient(ctx context.Context, hostURL string) (*govmomi.Client, error) {
+	if o.Client != nil {
+		return o.Client, nil
+	}
+
+	u, err := soap.ParseURL(parseVCHost(hostURL))
 	if err != nil {
 		return nil, err
 	}
-	return govmomi.NewClient(ctx, u, o.insecure)
+	o.processCredential(u)
+	o.url = u
+	client, err := govmomi.NewClient(ctx, o.url, o.insecure)
+	if err != nil {
+		return client, err
+	}
+	o.Client = client
+	return o.Client, nil
+}
+
+func (o *ClientOption) processCredential(u *url.URL) error {
+	// consider secret, and env for credentials
+
+	envUsername := o.getCredential().getUsername()
+	envPassword := o.getCredential().getPassword()
+	// Override username if provided
+	if envUsername != "" {
+		var password string
+		var ok bool
+
+		if u.User != nil {
+			password, ok = u.User.Password()
+		}
+
+		if ok {
+			u.User = url.UserPassword(envUsername, password)
+		} else {
+			u.User = url.User(envUsername)
+		}
+	}
+
+	// Override password if provided
+	if envPassword != "" {
+		var username string
+
+		if u.User != nil {
+			username = u.User.Username()
+		}
+
+		u.User = url.UserPassword(username, envPassword)
+	}
+	return nil
+}
+
+func (o *ClientOption) GetClient() (*vim25.Client, error) {
+	if o.Client.Client != nil {
+		return o.Client.Client, nil
+	}
+	return nil, nil
+}
+
+func (o *ClientOption) Userinfo() *url.Userinfo {
+	return o.url.User
+}
+
+func (o *ClientOption) getCredential() Credential {
+	return o.credential
+}
+
+type Credential struct {
+	username string
+	password string
+	cert     string
+	Secrete  VCCMSecret
+}
+
+type VCCMSecret struct {
+	Name string
+	Data map[string]string
+}
+
+func (o *ClientOption) LoadCredential(username, password, cert string) {
+	c := Credential{}
+	c.setUsername(username)
+	c.setPassowrd(password)
+	c.setCertificate(cert)
+	// TODO: (fanz) Secret of Credential
+	o.credential = c
+}
+
+func (c Credential) setPassowrd(p string) {
```

Comment:
`setPassowrd` -> `setPassword`

Reply:
Why the getters and setters here, rather than just use the fields directly?

Reply:
Removed setters and getters.